### DISCUSSION TITLE
Janet Emacs Trial Kit

### DISCUSSION THREAD

**Discussion by sogaiu** [janet-lang/janet#1300] [Show and tell]
I've put together a [repository](https://github.com/sogaiu/janet-emacs-trial-kit) that should make it relatively easy to try out various Janet support features in recent Emacsen (>= 29.1) -- without messing with one's existing Emacs setup.

![janet-emacs-trial-kit-linux](https://github.com/janet-lang/janet/assets/33044872/f07578ac-7a96-4692-ab9b-537a03a7c27b)

The Emacs setup provides:

* Basic Janet code handling (e.g. syntax highlighting and indentation) via [janet-ts-mode](https://github.com/sogaiu/janet-ts-mode)
* REPL interaction via [ajrepl](https://github.com/sogaiu/ajrepl)
* Linting via [flycheck-janet](https://github.com/sogaiu/flycheck-janet)
* Rainbow delimiters
* Dark theme

All without having to:

* Manually fetch certain dependencies
* Change your existing Emacs configuration
* Figure out how to configure Emacs for some Janet support

Assuming basic prerequisites (appropriate c compiler, emacs (>= 29.1 with tree-sitter support) git, janet ~~and jpm~~), getting started should be a matter of:

```
git clone https://github.com/sogaiu/janet-emacs-trial-kit
cd janet-emacs-trial-kit
janet jetk
```

...plus a bit of patience as various bits are fetched, compiled, etc. :)

Please see the [repository](https://github.com/sogaiu/janet-emacs-trial-kit) for more details (including further details on getting this to work on Windows).

**Comment by sogaiu** [comment]
I've updated the instructions above as `jpm` is no longer a prerequisite and invocation has changed.

I think this should make it a bit easier to try things out.

---

A similar change is planned for [janet-neovim-trial-kit](https://github.com/sogaiu/janet-neovim-trial-kit) -- just waiting on [an update to Conjure to support an alternative interaction method that doesn't need netrepl](https://github.com/Olical/conjure/pull/536).

_Update_: the change has been made.

**Comment by dawranliou** [comment]
Thank you for the great work providing all those Emacs packages for the Janet ecosystem!  As someone who's a relatively new Janet, what are the major differences between https://github.com/sogaiu/ajrepl vs https://github.com/velkyel/inf-janet?

**Reply by sogaiu** [reply]
It's been a while since I've looked at inf-janet in detail but I did take some notes about it and ijanet at one point so I'll make an initial attempt at a summary.

Looking over them now, it appears that I was under the impression that:

* form recognition might not handle certain Janet constructs (e.g. `@[1 2 3]` or `@{:a 1 :b 2}`)
* evaluation results didn't appear to show up in the repl buffer
* lacking menu (for (re)discoverability)

These were things that I wanted personally so I tried to address them via ajrepl, but perhaps some or all of them are not that high priority for other folks.  It's also possible things have changed or I was mistaken when examining before (^^;

**Reply by dawranliou** [reply]
Thanks for the summary!  I'll give ajrepl a shot because I also dig its name. :)

**Comment by ghost** [comment]
Awesome awesome awesome! I know I'm late to the party but could you please also make this a [Spacemacs](spacemacs.org) layer, if it's not too much work? (I'd help but I don't have a dev machine atm and I probably couldn't commit to ongoing maintenance due to mental health, so I'm definitely not making any demands.)

**Reply by sogaiu** [reply]
I haven't had much experience with Spacemacs recently, but the other day I noticed [this](https://github.com/5thWall/spacemacs-janet-layer).  May be that's of some help?
